app/src/main/kotlin/compiler/intermediate_form/Linearization.kt

```diff
@@ -1,10 +1,111 @@
 package compiler.intermediate_form
 
 import compiler.common.intermediate_form.Covering
+import compiler.common.reference_collections.referenceHashMapOf
+import compiler.intermediate_form.Instruction.UnconditionalJumpInstruction.Jmp
 
 object Linearization {
     fun linearize(
         cfg: ControlFlowGraph,
         covering: Covering,
-    ): List<Asmable> = TODO()
+    ): List<Asmable> {
+        if (cfg.entryTreeRoot == null)
+            return emptyList()
+
+        val instructions = mutableListOf<Asmable>()
+        val labels = referenceHashMapOf<IFTNode, String>()
+        val endLabel = "end"
+        val usedLabels = mutableSetOf<String>()
+
+        fun assignLabel(node: IFTNode): String {
+            val label = "_" + labels.size
+            labels[node] = label
+            return label
+        }
+
+        fun addLabel(label: String) {
+            instructions.add(Label(label))
+        }
+
+        fun dfs(node: IFTNode, nextLabel: String) {
+            addLabel(labels[node] ?: assignLabel(node))
+
+            fun addUncoditional() {
+                instructions.addAll(covering.coverUnconditional(node))
+            }
+
+            fun addConditional(whenTrue: Boolean, targetLabel: String) {
```

Comment:
We use "link type" in the context of CFGs to mean either "unconditional", "conditional when the result is true", or "conditional when the result false". Since the name of the function is already `addConditional`, changing the parameter name to `linkType` would be confusing.